### ISSUE TITLE
Spurious JSC_MISSING_REQUIRE_IN_PROVIDES_FILE when importing @interface type via goog.requireType

### ISSUE THREAD

**Issue by cpcallen** [google/closure-compiler#3851]
The following code, [when compiled with `CHECK_REQUIRES`, produces a `JSC_MISSING_REQUIRE_IN_PROVIDES_FILE` warning](https://closure-compiler-debugger.appspot.com/#input0%3Dgoog.provide('Interface')%253B%250A%250A%252F**%2520%2540interface%2520*%252F%250AInterface%2520%253D%2520function()%2520%257B%257D%253B%250A%26input1%3Dgoog.requireType('Interface')%253B%250A%250A%252F**%2520%2540constructor%2520%2540implements%2520%257BInterface%257D%2520*%252F%250Afunction%2520Implementation()%2520%257B%257D%250A%26conformanceConfig%26externs%26refasterjs-template%26CHECK_TYPES%3Dtrue%26REWRITE_MODULES_BEFORE_TYPECHECKING%3Dtrue%26CHECK_REQUIRES%3Dtrue%26CLOSURE_PASS%3Dtrue%26PRESERVE_TYPE_ANNOTATIONS%3Dtrue%26PRETTY_PRINT%3Dtrue):

Parent:
```JS
goog.provide('Interface');

/** @interface */
Interface = function() {};
```

Child:
```
goog.requireType('Interface');

/** @constructor @implements {Interface} */
function Implementation() {}
```

Warnings:
```
input1:3:30: WARNING - [JSC_MISSING_REQUIRE_IN_PROVIDES_FILE] 'Interface' references a namespace which was not required by this file.
Please add a goog.require.
  3| /** @constructor @implements {Interface} */
                                   ^^^^^^^^^

0 error(s), 1 warning(s), 55.5% typed
```

I believe this warning is spurious, because [it does not occur if you use `goog.require` instead of `goog.requireType`](https://closure-compiler-debugger.appspot.com/#input0%3Dgoog.provide('Interface')%253B%250A%250A%252F**%2520%2540interface%2520*%252F%250AInterface%2520%253D%2520function()%2520%257B%257D%253B%26input1%3Dgoog.require('Interface')%253B%250A%250A%252F**%2520%2540constructor%2520%2540implements%2520%257BInterface%257D%2520*%252F%250Afunction%2520Implementation()%2520%257B%257D%250A%26conformanceConfig%26externs%26refasterjs-template%26CHECK_TYPES%3Dtrue%26REWRITE_MODULES_BEFORE_TYPECHECKING%3Dtrue%26CHECK_REQUIRES%3Dtrue%26CLOSURE_PASS%3Dtrue%26PRESERVE_TYPE_ANNOTATIONS%3Dtrue%26PRETTY_PRINT%3Dtrue), even though the import is used only for type annotation, and [it does not occur if you use `goog.module` instead of `goog.provide`](https://closure-compiler-debugger.appspot.com/#input0%3Dgoog.module('Interface')%253B%250A%250A%252F**%2520%2540interface%2520*%252F%250Aconst%2520Interface%2520%253D%2520function()%2520%257B%257D%253B%250Aexports%2520%253D%2520Interface%253B%250A%26input1%3Dgoog.module('main')%253B%250Aconst%2520Interface%2520%253D%2520goog.requireType('Interface')%253B%250A%250A%252F**%2520%2540constructor%2520%2540implements%2520%257BInterface%257D%2520*%252F%250Afunction%2520Implementation()%2520%257B%257D%253B%250A%26conformanceConfig%26externs%26refasterjs-template%26CHECK_TYPES%3Dtrue%26REWRITE_MODULES_BEFORE_TYPECHECKING%3Dtrue%26CHECK_REQUIRES%3Dtrue%26CLOSURE_PASS%3Dtrue%26PRESERVE_TYPE_ANNOTATIONS%3Dtrue%26PRETTY_PRINT%3Dtrue).

**Comment by h-joo** [comment]
@lauraharker I think I need your look on this one. Do you think this is a valid bug?

**Comment by lauraharker** [comment]
This was an intentional choice when the CheckMissingRequires pass was written. The problem was that the typechecker doesn't support `@implements`/`@extends` of a weak require very well. See also https://github.com/google/closure-compiler/issues/3583.

Here's the code in question: https://github.com/google/closure-compiler/blob/78ce1e53ae22900b5f7d5dad19a34babf1273763/src/com/google/javascript/jscomp/CheckMissingRequires.java#L146-L153.

One fix might be making the error message more explicit about why this error pops up for requireTyped names.

**Comment by cpcallen** [comment]
Ahh, interesting.  Am I correct in understanding that the bug is not the error when using `goog.provides`, but the lack of error when using `goog.module`?  If so, that is quite surprising, and violates my (I think not entirely unreasonable) assumption of a rule that that an import which is mentioned only in comments (i.e., used only as a type) need only be `requireType`d rather than `require`d.  It seems especially egregious that it is interface types which need be `require`d, since they will never be mentioned in the source other than in comments as a type.

Is the requirement to `require` interface types documented somewhere other than in `CheckMissingRequires.java`?

**Comment by concavelenz** [comment]
You should alway see the error when a interface is implemented if it isn't goog.required regardless of whether the original source is a goog.provide or a goog.module.

**Comment by cpcallen** [comment]
I've updated the title and description to better match my updated understanding of the desired/expected behaviour based on the discussion above.
